Support for multiple signature scheme for BEEFY primitves

drskalman opened the pull request:
This is a subset of the changes introduced by Pull request #13311 which tends to enable BEEFY to support more than one signature scheme at the time in order to accommodate various bridged chains. This is only the changes needed in `primitives/consensus/beefy` while keeping the current client backward compatible in order to facilitate reviewing and merging.

polkadot companion: https://github.com/paritytech/polkadot/pull/7572

cumulus companion: https://github.com/paritytech/cumulus/pull/2962

acatangiu commented on primitives/consensus/beefy/src/lib.rs:
```diff
@@ -303,15 +322,14 @@ impl OpaqueKeyOwnershipProof {
-	/// API necessary for BEEFY voters.
-	#[api_version(2)]
-	pub trait BeefyApi
+	/// API necessary for BEEFY voters with only ECDSA key.
+	pub trait BeefyApi<AuthorityId> where AuthorityId : Encode + Decode
```
Is this some intermediary step before supporting full API using other than ECDSA?

With this new code, the BEEFY runtime API is generic on `AuthorityId`, but the equivocation functions are using hardcoded `ecdsa_crypto::AuthorityId` which is inconsistent/incorrect when generic `AuthorityId` is other than ECDSA.

What is keeping us from making full api generic and consistent?

drskalman replied:
Resolved according to our discussion.

acatangiu reviewed (commented):
Everything looks good, except single question below.

davxy commented on primitives/consensus/beefy/Cargo.toml:
```diff
@@ -24,10 +24,14 @@ sp-runtime = { version = "24.0.0", default-features = false, path = "../../runti
 array-bytes = "4.1"
 sp-keystore = { version = "0.27.0", path = "../../keystore" }
+w3f-bls = { version = "0.1.3", features = ["std"]} 
+hex = { version = "0.4", default-features = false}
+hex-literal = "0.3.4"
```
please use `array_bytes` crate for the same goal (we're trying to reduce dependencies offering duplicated functionality)

davxy replied:
This applies to `sp_core::bls` as well

drskalman replied:
Removed and replaced in `sp_core::bls` as well.

davxy commented on primitives/consensus/beefy/src/lib.rs:
```diff
@@ -303,15 +322,14 @@ impl OpaqueKeyOwnershipProof {
-	#[api_version(2)]
-	pub trait BeefyApi
+	/// API necessary for BEEFY voters with only ECDSA key.
+	pub trait BeefyApi<AuthorityId> where AuthorityId : Encode + Decode
```
silly nit (feel free to ignore)

```suggestion
	pub trait BeefyApi<AuthorityId> where AuthorityId : Codec
```

davxy commented on primitives/consensus/beefy/src/witness.rs:
```diff
@@ -80,15 +83,36 @@ mod tests {
+
+	#[cfg(feature = "bls-experimental")]
+	#[derive(Clone, Debug, PartialEq, codec::Encode, codec::Decode)]
+	struct ECDSABLSSignaturePair(ecdsa_crypto::Signature, BLSSignature);
```
Please use camel case for all novel type names (also for acronyms like ECDSA and BLS).
This is a de-facto standard practice in Substrate and Rust in general (see https://rust-lang.github.io/api-guidelines/naming.html)

```suggestion
	struct EcdsaBlsSignaturePair(ecdsa_crypto::Signature, BlsSignature);
```

drskalman replied:
CamelCased!

davxy commented:
Probably at some point you'll need to embed this key type in the keys used by some node (in SessionKeys [here[(https://github.com/paritytech/substrate/blob/ae1a608c91a5da441a0ee7c26a4d5d410713580d/bin/node/runtime/src/lib.rs#L512-L519)).

This is used to generate new keys "on demand" via the `sp_session::SessionKeys` interface or in the genesis build for the testnet node.

BTW, to add a key type there you need to implement the `RuntimePublic` trait for your application specific public key.
Is not required to implement the full set of functions (as you don't need all of them) you can implement only the `generate_key`.

As a reference see the bandersnatch-vrf PR here: https://github.com/paritytech/substrate/blob/bdd3df4ec618de1010ad7c5c808666771127ed36/primitives/application-crypto/src/bandersnatch.rs#L32-L57

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3046381

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3046382

drskalman commented:
> BTW, to add a key type there you need to implement the `RuntimePublic` trait for your application specific public key.

Hmm, I [did](https://github.com/paritytech/substrate/pull/13618/commits/b84521016c6ae96a50ba8a339f53c704c0c0f201#diff-0ac28479b887d29368c53efdf6c870e9b8b67290bc4614d7420aa4021cfff587R40) but you removed them for some reason. I can bring them back if you think they are OK.

davxy commented:
> > BTW, to add a key type there you need to implement the `RuntimePublic` trait for your application specific public key.
> 
> Hmm, I [did](https://github.com/paritytech/substrate/pull/13618/commits/b84521016c6ae96a50ba8a339f53c704c0c0f201#diff-0ac28479b887d29368c53efdf6c870e9b8b67290bc4614d7420aa4021cfff587R40) but you removed them for some reason. I can bring them back if you think they are OK.

@drskalman we don't require to implement the full set of methods because most of them requires host functions. 
An is preferable to not add host functions if are not really-really required.

Unfortunately, working on analoguous code for introducing bandersnatch-vrf, I noticed that if you want to use the key as part of the session keys set you require to implement **at least** the `generate_pair` because it is used to do a couple of things.

Now, as you can see from the snip I've shared I just implemented the `generate_pair` and provided a dummy implementation for the others (makes no sense introducing a HF that is never used).

Before adding it, just let me check a couple of things to see if we can instead fix this requirement in other way

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3073338

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3073341

drskalman commented:
@davxy : It seems that `runtime_interface` is failing to derive `PassBy` using your custom implementation of `PassByInner`:
```
    |   the trait `PassBy` is not implemented for `sp_core::bls::Public<w3f_bls::engine::TinyBLS<ark_ec::models::bls12::Bls12<ark_bls12_377::curves::Config>,
 ark_bls12_377::curves::Config>>`                                                                                                                     
``` 

Should we also manually implement `PassBy` for Public And Signature in `core/src/bls.rs`?

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3183070

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3183071

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3204981

davxy commented on primitives/consensus/beefy/src/witness.rs:
```diff
@@ -57,83 +60,191 @@ impl<TBlockNumber, TMerkleRoot> SignedCommitmentWitness<TBlockNumber, TMerkleRoo
+	///BLS signatures
+	#[cfg(feature = "bls-experimental")]
+	fn mock_bls_signatures() -> (BlsSignature, BlsSignature) {
+		let store: KeystorePtr = MemoryKeystore::new().into();
+
+		let alice = sp_core::bls::Pair::from_string("//Alice", None).unwrap();
+		store.insert(BEEFY_KEY_TYPE, "//Alice", alice.public().as_ref()).unwrap();
```
```suggestion
		let alice = sp_core::bls::Pair::from_string("//Alice", None).unwrap();
```
You don't need a keystore for this test.

drskalman replied:
Done.

davxy commented on frame/beefy/src/equivocation.rs:
```diff
@@ -41,7 +41,8 @@ use frame_support::{
 use frame_system::pallet_prelude::BlockNumberFor;
 use log::{error, info};
-use sp_consensus_beefy::{EquivocationProof, ValidatorSetId, KEY_TYPE};
+use sp_application_crypto::key_types::BEEFY as BEEFY_KEY_TYPE;
```
NIT. You can remove the extra `sp-application-crypto` dependency by re-exporting the `BEEFY` key type in `sp_consensus_beefy` (as it has been done from [babe](https://github.com/paritytech/substrate/blob/cfe2432fdba3ff498b3c867c18bb4c25d0625a74/primitives/consensus/babe/src/lib.rs#L40) and [grandpa](https://github.com/paritytech/substrate/blob/cfe2432fdba3ff498b3c867c18bb4c25d0625a74/primitives/consensus/grandpa/src/lib.rs#L44) engines)

davxy reviewed (commented):
I went through the PR and generally looks good.

I left some comments here and there where IMO the code can be improved (mostly nits).

BTW looks like the `client` part  is not generic over the `AuthorityId` yet right?
Maybe you can get some inspiration from the AURA code

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3216264

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3216263

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3216313

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3216314

drskalman commented:
> BTW looks like the `client` part is not generic over the `AuthorityId` yet right? Maybe you can get some inspiration from the AURA code

Yes, that is @acatangiu wish to break the [original PR](https://github.com/paritytech/substrate/pull/13311) in three parts for an easier more-manageable review.

drskalman commented on client/consensus/beefy/src/communication/gossip.rs:
```diff
@@ -41,7 +41,7 @@ use crate::{
 };
 use sp_consensus_beefy::{
-	crypto::{AuthorityId, Signature},
+	ecdsa_crypto::{AuthorityId, Signature},
```
Except in our use case, it is not always a `P:Public` type. We are currently looking at a pair of public keys for example: (ECDSA Public key, BLS Public Key). And the signature is a pair of signatures. The idea is to verify ECDSA signature during gossip and go back and slash if the aggregated BLS signatures didn't verify. So we need a deeper level of abstraction here than AURA.

So following @burdges suggestion, I implemented it like [this](https://github.com/paritytech/substrate/pull/13311/files#diff-c29dfa7a41414cb5dcb9fc5dcf7aab90c60ec743374709c5a75a8e94f4517cd5R257).

But I'm open if you think we can do it in a better way.

drskalman commented on primitives/consensus/beefy/src/lib.rs:
```diff
@@ -104,6 +98,28 @@ pub mod crypto {
+
+	/// Signature for a BEEFY authority using BLS as its crypto.
+	pub type AuthoritySignature = Signature;
+}
```
Ditto. And verifying (ECDSA, BLS) signature is different as we only verifying ECDSA in gossip.

But I wonder if it is better the way it is done now in #13311 and make the BEEFY keystore generic over AuthId or we should actually define `pub mod ecdsa_n_bls_crypto` in `primitives/consensu/beefy` instead.

What do you think?

drskalman replied:
My suggestion does not work out of the box because `type AuthorityId = (ecdsa_crypto::Public, bls_crypto::Public)` does not implement `RuntimeAppPublic` But we still can have mod for this type of `AuthorityId` and use it in client. We'll resolve this restricton in future improvement so we can implement BeefyAuthorityId for `(ecdsa_crypto::Public, bls_crypto::Public)`

drskalman replied:
> But can't we just call the Pair::verify so that we implement the BeefyAuthorityId over a generic AuthorityId<P> where P: Pair?

The reason they do not call `Pair::verify` beause they want to have control over the hasher which I believe Pair::verify doesn't give.

drskalman replied:
I also have to re-introduce `w3f-bls` as optional dependancy to implement `BeefyAuthorityId` for `bls_crypto::AuthorityId` because we do not have a host call for verifying bls signatures.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3218753

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3218754

drskalman commented on primitives/consensus/beefy/src/witness.rs:
```diff
@@ -57,83 +60,191 @@ impl<TBlockNumber, TMerkleRoot> SignedCommitmentWitness<TBlockNumber, TMerkleRoo
 
 	// The mock signatures are equivalent to the ones produced by the BEEFY keystore
-	fn mock_signatures() -> (crypto::Signature, crypto::Signature) {
+	fn mock_ecdsa_signatures() -> (ecdsa_crypto::Signature, ecdsa_crypto::Signature) {
```
Perhaps because they are going to ue the keystore further on in the client/beefy/BeefyKeystore, perhaps they wanted to use the exact mock signature they are going to produce later on? In this way they are testing the keystare functionality at the same time. That is what I understand from the comment.

What do you think? If you still think we shouldn't test keystore involvement here, I would remove it.

davxy replied:
I think that the keystore should be tested by the keystore tests.
Unit tests should reduce the test surface as most as possible

drskalman replied:
Keystore removed from `dev-dependencies`

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3244564

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3245047

drskalman commented on primitives/consensus/beefy/src/commitment.rs:
```diff
@@ -249,30 +249,75 @@ impl<N, S> From<SignedCommitment<N, S>> for VersionedFinalityProof<N, S> {
+	///generates mock aggregatable bls signature for generating test commitment
+	///BLS signatures
+	#[cfg(feature = "bls-experimental")]
+	fn mock_bls_signatures() -> (BlsSignature, BlsSignature) {
```
Done

drskalman commented on primitives/consensus/beefy/src/lib.rs:
```diff
@@ -368,12 +385,12 @@ mod tests {
 	fn beefy_verify_works() {
 		let msg = &b"test-message"[..];
-		let (pair, _) = crypto::Pair::generate();
+		let (pair, _) = ecdsa_crypto::Pair::generate();
```
https://github.com/paritytech/substrate/pull/14373/files#diff-22eeed1a8484727d7e5dbe89a81390010c9cccc416c45795c09a9e4ee8fd0948R463

drskalman commented on primitives/consensus/beefy/src/witness.rs:
```diff
@@ -57,83 +60,191 @@ impl<TBlockNumber, TMerkleRoot> SignedCommitmentWitness<TBlockNumber, TMerkleRoo
 	use codec::Decode;
 
-	use crate::{crypto, known_payloads, Payload, KEY_TYPE};
+	use crate::{ecdsa_crypto, known_payloads, Payload};
```
Done.

drskalman commented:
@davxy I believed I addressed all your comment, except I had to add `w3f-bls` dependency because we do not have a host function to verify `bls` signatures.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3245208

davxy commented on Cargo.lock:
I'm not sure if we want to bulk update all the substrate dependencies via this PR.

Maybe is better if you pick the `Cargo.lock` from master and add to this PR only the modifications stemming from a plain `cargo build` (without issuing a `cargo update`)

davxy commented:
> @davxy I believed I addressed all your comment, except I had to add w3f-bls dependency because we do not have a host function to verify bls signatures.

@drskalman I added a suggestion that allows to remove direct dep from w3f-bls.

Some more nits

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3271430

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3271431

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3280457

drskalman commented on client/consensus/beefy/src/keystore.rs:
hmm, I'm not sure, You know the structure of Keystore better. But my guess was that, it was some safe coding practice. That because the pointer could be invalid or something you need the option of setting it to be not. There was a test for checking BeefyKestore initiated with None and I couldn't pass it after I changed the keystore structure and so I deleted it in my (closed) PR but I can't find it somehow :-/ Maybe that test sheds some ligh.

drskalman replied:
In the current code:
https://github.com/paritytech/substrate/blob/master/client/consensus/beefy/src/keystore.rs#L255

drskalman replied:
Maybe if you are not a validator then you can run beefy with None keystore.

drskalman commented:
> @drskalman I added a suggestion that allows to remove direct dep from w3f-bls.
> 
> Some more nits

I think I addressed all the new ones.

andresilva commented on primitives/consensus/beefy/src/lib.rs:
```diff
@@ -104,6 +101,44 @@ pub mod crypto {
 
+/// BEEFY cryptographic types for BLS crypto
+///
+/// This module basically introduces three crypto types:
```
```suggestion
/// This module basically introduces four crypto types:
```

andresilva commented on primitives/consensus/beefy/src/lib.rs:
```diff
@@ -63,23 +63,20 @@ pub trait BeefyAuthorityId<MsgHash: Hash>: RuntimeAppPublic {
-/// BEEFY cryptographic types
+/// BEEFY cryptographic types for ECDSA crypto
 ///
 /// This module basically introduces three crypto types:
```
```suggestion
/// This module basically introduces four crypto types:
```

andresilva commented on primitives/consensus/beefy/src/lib.rs:
```diff
@@ -63,23 +63,20 @@ pub trait BeefyAuthorityId<MsgHash: Hash>: RuntimeAppPublic {
-/// - `crypto::Signature`
+/// - `ecdsa_crypto::Pair`
+/// - `ecdsa_crypto::Public`
+/// - `ecdsa_crypto::Signature`
```
```suggestion
/// - `ecdsa_crypto::Signature`
/// - `ecdsa_crypto::AuthorityId`
```

andresilva commented on primitives/consensus/beefy/src/witness.rs:
```diff
@@ -57,89 +60,183 @@ impl<TBlockNumber, TMerkleRoot> SignedCommitmentWitness<TBlockNumber, TMerkleRoo
 
 		// when
+		let (witness, _signatures) =
+		        //from signed take a function as the aggregator 
+			TestBlsSignedCommitmentWitness::from_signed::<_, _>(signed, |sigs| {
+			    //we are going to aggregate the signatures here
+			    let mut aggregatedsigs: SignatureAggregatorAssumingPoP<TinyBLS377> =
+				SignatureAggregatorAssumingPoP::new(Message::new(b"", b"mock payload"));
+
+			    for sig in sigs {
+				match sig {
+				    Some(sig) =>
+				    {
+					let serialized_sig : Vec<u8> = (*sig.1).to_vec();
+					aggregatedsigs.add_signature(
+					    &w3f_bls::Signature::<TinyBLS377>::from_bytes(
+						serialized_sig.as_slice()
+					     ).unwrap());
+					    },
+				    None => (),
+				}
+			    }
+			    (&aggregatedsigs).signature().to_bytes()
+			});
```
```suggestion
			// from signed take a function as the aggregator 
			TestBlsSignedCommitmentWitness::from_signed::<_, _>(signed, |sigs| {
				// we are going to aggregate the signatures here
				let mut aggregatedsigs: SignatureAggregatorAssumingPoP<TinyBLS377> =
					SignatureAggregatorAssumingPoP::new(Message::new(b"", b"mock payload"));

				for sig in sigs {
					match sig {
						Some(sig) => {
							let serialized_sig : Vec<u8> = (*sig.1).to_vec();
							aggregatedsigs.add_signature(
								&w3f_bls::Signature::<TinyBLS377>::from_bytes(
									serialized_sig.as_slice()
								).unwrap()
							);
						},
						None => (),
					}
				}
				(&aggregatedsigs).signature().to_bytes()
			});
```

andresilva replied:
Just formatting changes.

andresilva reviewed (approved):
lgtm. just minor nits

Lederstrumpf commented on primitives/application-crypto/src/bls377.rs:
```diff
@@ -26,3 +27,30 @@ mod app {
+	}
+
+	fn generate_pair(key_type: KeyTypeId, seed: Option<Vec<u8>>) -> Self {
+		sp_io::crypto::bls377_generate(key_type, seed)
```
to my understanding, the consensus so far is to delay adding host functionality as long as possible, and instead use bls libraries directly within the runtime.
so will all other implementations here remain dummies and use runtime functions instead?
I ask because this does in fact expose bls377 key generation from the host, but this is not useful unless `sign` is implemented too. Else, one might as well keep `generate_pair` a dummy stub too.

davxy replied:
Unfortunately there is another code path that requires `generate_pair`.

There are a couple of places ([here](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/client/rpc/src/author/mod.rs#L129-L132) and [here](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/primitives/session/src/lib.rs#L132-L134)) outside the runtime which rely on the runtime (for convenience) to generate the pairs for each element in the `SessionKeys` ([this](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/bin/node-template/runtime/src/lib.rs#L87-L92) ).

For example, for the node-template:
- The runtime API implementation to generate a keypair for each element in `SessionKeys` is [this](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/bin/node-template/runtime/src/lib.rs#L418-L420) 
- `generate_session_keys` calls `generate_pair` for each `RuntimeAppPublic` in the `SessionKeys` tuple. See [here](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/primitives/runtime/src/traits.rs#L2010-L2021)

drskalman replied:
See [this comment](https://github.com/paritytech/substrate/pull/14373#issuecomment-1602043252) from @davxy. Keys are generated by the session pallet in the Runtime and so we have no easy way to customize this (and you need good randomness for that so it is not even a good idea to avoid `std` for that) . In contrast, `sign` will almost never be used in the Runtime so that should be Ok. For `verify`, because it is (currently) only being used in BEEFY pallet and we are free to have our own implementation of `verify` there therefore we could implement  the whole verify function in the runtime without involving a new host function, as it is done [here](https://github.com/w3f/substrate/blob/skalman-multiple-signature-scheme-beefy-primitves/primitives/consensus/beefy/src/lib.rs#L139C1-L139C1).

If someone else wants to use bls12-377 crypto for something else and does not know that verify is dummy then they'll be annoyed that all their signatures fail to verify till they check what is actually happening under the hood.

drskalman replied:
on the bright side we have the `bls-experimental` feature flag for now :-)

Lederstrumpf replied:
ack, thanks your explanations & pointing me to those uses of `generate_pair` - also don't see a workaround for this.

burdges replied:
> There are a couple of places ([here](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/client/rpc/src/author/mod.rs#L129-L132) and [here](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/primitives/session/src/lib.rs#L132-L134)) outside the runtime which rely on the runtime (for convenience) to generate the pairs for each element in the SessionKeys ([this](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/bin/node-template/runtime/src/lib.rs#L87-L92) ).

@davxy  Is this so that we can deprecate old session keys by no longer doing the invocation?  Or some other reason?

davxy replied:
I'm not aware of the original rationale of such design decision, but if I have to guess is to have some *cryptographic agility*. In this way we can add/remove pallets to the runtime which require a session key without worrying too much since the runtime is the only in charge of directly altering the keys set structure

Lederstrumpf reviewed (approved):
lgtm - one comment re. consistency on path forward wrt. host functions.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3313025

davxy commented:
bot merge

davxy commented:
@drskalman apparently you have to enable "allow edits from maintainers" in your polkadot PR

ironoa commented:
> @drskalman apparently you have to enable "allow edits from maintainers" in your polkadot PR

there seems to be a known GH issue that still hasn't been solved, which prevents "allow edits from maintainers" to be engaged on our end: https://github.com/orgs/community/discussions/5634

Have you ever managed to bypass this problem ?

davxy commented:
> > @drskalman apparently you have to enable "allow edits from maintainers" in your polkadot PR
> 
> there seems to be a known GH issue that still hasn't been solved, which prevents "allow edits from maintainers" to be engaged on our end: https://github.com/orgs/community/discussions/5634
> 
> Have you ever managed to bypass this problem ?

cc @paritytech/release-engineering

davxy commented:
bot merge
